### ISSUE TITLE
OneTrainer adding torch compile perhaps we can also have? free speed up

### ISSUE THREAD

**Issue by FurkanGozukara** [kohya-ss/musubi-tuner#716]
here their pull request : https://github.com/Nerogar/OneTrainer/pull/1083

According to Codex here a plan @kohya-ss 

### Overview
This document outlines the steps needed to bring `torch.compile` optimizations to Musubi Tuner’s Qwen-Image training workflow. Each phase builds on the previous one so we can ship incremental gains while preserving stability.

---

### Phase 1 · Base Toggle & Happy-Path Compile
- **CLI support**
  - Add `--compile` and optional knobs (`--compile-backend`, `--compile-mode`, `--compile-dynamic`, `--compile-fullgraph`) to `qwen_image_train.py`.
  - Persist the flag in saved configs / resume metadata for reproducibility.
- **Where to compile**
  - After `self.load_transformer(...)` and any LoRA or quantization merges, but before `accelerator.prepare`, wrap every block in `transformer.transformer_blocks` with `torch.compile`.
  - Default settings: `backend="inductor"`, `mode="reduce-overhead"`, `dynamic=False`, `fullgraph=True`.
  - Set `torch._dynamo.config.cache_size_limit = 8192` to reduce recompiles.
- **Initial safeguards**
  - Skip compilation (with a log message) when:
    - `args.blocks_to_swap > 0`
    - `args.gradient_checkpointing_cpu_offload`
    - `args.fp8_scaled`

---

### Phase 2 · Gradient Checkpointing & CPU Offload Compatibility
- **Checkpoint wrappers**
  - Introduce wrappers (similar to OneTrainer’s `CheckpointLayer`) that integrate gradient checkpointing with compiled modules instead of relying solely on `requires_grad_`.
  - For CPU offload, compile the underlying block (`orig_module.compile(fullgraph=True)`) and let the wrapper manage activation movement.
- **Offload-friendly wrapper**
  - Ensure `create_cpu_offloading_wrapper` stays Dynamo-friendly by keeping the wrapped function graphable.
- **Guard removal**
  - Drop the `gradient_checkpointing_cpu_offload` compile guard once the new wrapper passes smoke tests.

---

### Phase 3 · Block Swapping Support
- **Compiled blocks + swapper**
  - Update `ModelOffloader` to handle Dynamo’s `OptimizedModule` (resolve wrapped attributes as needed).
- **Compilation order**
  - Compile each block before swap hooks take over so the compiled module is what gets swapped.
- **Guard removal**
  - When swapping + compile passes testing, remove the `blocks_to_swap` guard.

---

### Phase 4 · Quantization & Mixed Precision
- **Compile after quantization**
  - Ensure all FP8/INT8 replacements happen before `torch.compile` executes so the compiler sees final operator graphs.
- **FP8 validation**
  - Test FP8-base and standard runs; if FP8 kernels cause Dynamo bailouts, patch those kernels rather than disabling compile.

---

### Phase 5 · Inference & Shared Utilities
- Reuse the new compile switches in inference scripts (`qwen_image_generate_image`, video sampling) for consistent speedups.
- Optionally centralize compile helpers in a shared utility module to make adoption by other architectures easier.

---

### Phase 6 · Validation & Observability
- **Benchmarks**
  - Add short training smoke tests (e.g., 10 steps) for both text-to-image and edit modes, with/without gradient checkpointing, to compare throughput and surface graph breaks.
- **Logging & metadata**
  - Log compile status, backend, and mode so users can confirm it’s active.
  - Store compile settings in metadata exported with checkpoints.
- **Docs**
  - Update public docs/README with guidance on enabling compile, known incompatibilities, and troubleshooting tips.

---

### Next Steps
Start with Phase 1 to land a safe, opt-in compile path and measure gains. We can then iterate on Phases 2–4 to reach feature parity with OneTrainer. Let me know when you want code assistance for each phase.

**Comment by sdbds** [comment]
Does the current version of the compiler support dynamic resolution? I remember that dynamic resolution had many bugs on Windows, to the point where I rarely used it for image training (video training is more suitable)

**Comment by kohya-ss** [comment]
Thanks for raising this. I hadn't really considered torch.compile because it wasn't available on Windows.

I did a quick test with LoRA training for Qwen-Image on Windows, using a batch size of 1, without block swapping or fp8 scaling.

Musubi Tuner widely uses `--split_attn`, which speeds up attention by not using attention masks, so the input shape is different each time. Therefore, `dynamic=False` causes many compilations. Fortunately, `dynamic=True` seems to work in my testing.

With `mode="default", dynamic=True`, the training time was reduced by just under 25% (a 30% speedup), while with `dynamic=False` it was only about 10% (after compiled).

You can enable it by adding the following code to the following location:

https://github.com/kohya-ss/musubi-tuner/blob/85d047b99a3b4272da73addb0e047933a9b82624/src/musubi_tuner/qwen_image_train_network.py#L351
```
        # test: torch.compile
        logger.info("Compiling DiT model with torch.compile...")
        torch._dynamo.config.cache_size_limit = 32
        for i, block in enumerate(model.transformer_blocks):
            block = torch.compile(block, backend="inductor", mode="default", dynamic=True, fullgraph=False)
            model.transformer_blocks[i] = block

        return model
```

Also, in a quick test, `--fp8_scaled` worked, but `--blocks_to_swap` didn't.

**Comment by FurkanGozukara** [comment]
@kohya-ss awesome so you plan to add this as args? at least no block swap can enjoy for now

with block swap that be great 

also you mean speed up is 10% with `dynamic=True` ?  which we need for `--split_attn`

i am gonna test now

by the way i always train with single resolution so it is totally useful

**Comment by kohya-ss** [comment]
The effect was greater than I expected, so I would like to introduce it soon.

> also you mean speed up is 10% with `dynamic=True` ? which we need for `--split_attn`

Sorry, `dynamic=True` was more effective. I have corrected my previous comment.

**Comment by FurkanGozukara** [comment]
> The effect was greater than I expected, so I would like to introduce it soon.
> 
> > also you mean speed up is 10% with `dynamic=True` ? which we need for `--split_attn`
> 
> Sorry, `dynamic=True` was more effective. I have corrected my previous comment.

nice

by the way the impact is extremely significant

it reduced my VRAM usage, previously I had to do 7 block swap for 1328x1328 at LoRA rank 128, now it fits into RTX 5090 on Windows

```
        #return model
                # test: torch.compile
        logger.info("Compiling DiT model with torch.compile...")
        torch._dynamo.config.cache_size_limit = 32
        for i, block in enumerate(model.transformer_blocks):
            block = torch.compile(block, backend="inductor", mode="default", dynamic=False, fullgraph=False)
            model.transformer_blocks[i] = block

        return model
```

my best speed was like 5.9 second / it

now it is down to 4.60 second / it

we definitely need this with block swap @kohya-ss  and it works out of box amazing

this is a very dramatic speed up since it eliminated block swap for an empty 5090 

I will test lower LoRA rank 0 block swap difference too later with torch compile on and off - Qwen Base model training


 
<img width="3465" height="936" alt="Image" src="https://github.com/user-attachments/assets/49f47ddf-bd7e-47af-938e-c2bc060e0726" />

**Comment by AiInfluenceGr** [comment]
Great finding , congrats!

**Comment by FurkanGozukara** [comment]
ok i just did a test. qwen image lora , fp8 scaled, 64 lora rank, 1328x1328 pixels

Split Attention is working - all images are 1328x1328 and batch size = 1

with torch compile off

5.32 second / it 

<img width="1734" height="944" alt="Image" src="https://github.com/user-attachments/assets/d44b63a8-4799-4ce2-85b1-f9ab31194c6e" />

when it is on

it is 4.61 second / it

so speed gain is 13.3% with 0 trade off

<img width="1698" height="1430" alt="Image" src="https://github.com/user-attachments/assets/0127f583-daab-4e32-ba61-8384b7ea8255" />


with very basic torch compile on - immediately compiles for me

**Comment by AiInfluenceGr** [comment]
@kohya-ss thank you for the awareness. cant wait to train on 50% less time on windows!. (Im curently training with 18hours instead of 7 on linux :P )

**Comment by FurkanGozukara** [comment]
@kohya-ss OneTrainer pushed compile

they made fullgraph=True working on Windows as well with Block Swapping they have - so even better speed up

<img width="573" height="344" alt="Image" src="https://github.com/user-attachments/assets/2a1caedd-386d-42d7-a002-85506b23e6c7" />

here below their merged pull request

i hope we can get too, this will bring huge benefit

https://github.com/Nerogar/OneTrainer/pull/1114#event-20874039106

**Comment by FurkanGozukara** [comment]
I tested new merge of OneTrainer

When torch compile is off : 5.57 second / it

<img width="1731" height="909" alt="Image" src="https://github.com/user-attachments/assets/7a95c386-65b1-4b73-8a67-ef954f92aade" />

When torch compile is on : 4.77 second / it

<img width="1715" height="925" alt="Image" src="https://github.com/user-attachments/assets/eb7c8d15-bf3d-4ae8-be45-bc259f89984a" />


their speed gain is 14.3%

**Comment by kohya-ss** [comment]
We found that by excluding `nn.Linear` from compilation, `torch.compile` can be used in conjunction with block swap. Since the compilation effect of linear operations (matrix operations) itself should be quite low, we can expect some effectiveness even if `nn.Linear` is excluded from block swap.

**Comment by FurkanGozukara** [comment]
> We found that by excluding `nn.Linear` from compilation, `torch.compile` can be used in conjunction with block swap. Since the compilation effect of linear operations (matrix operations) itself should be quite low, we can expect some effectiveness even if `nn.Linear` is excluded from block swap.

Amazing

so many people looking forward to this

**Comment by Sana-A-E** [comment]
What's the current status on this?

**Comment by FurkanGozukara** [comment]
> What's the current status on this?

it is fully done working amazing
